src/lib/dlt_user.c

```diff
@@ -445,35 +453,37 @@ static DltReturnValue dlt_initialize_fifo_connection(void)
 
 DltReturnValue dlt_init(void)
 {
-    /* Compare 'dlt_user_initialised' to false. If equal, 'dlt_user_initialised' will be set to true.
-    Calls retruns true, if 'dlt_user_initialised' was false.
-    That way it's no problem, if two threads enter this function, because only the very first one will
-    pass fully. The other one will immediately return, because when it executes the atomic function
-    'dlt_user_initialised' will be for sure already set to true.
-    */
-    bool expected = false;
-    if (!(atomic_compare_exchange_strong(&dlt_user_initialised, &expected, true)))
+    /* process is exiting. Do not allocate new resources. */
+    if (dlt_user_freeing != 0) {
+        dlt_vlog(LOG_INFO, "%s logging disabled, process is exiting", __func__);
+        /* return negative value, to stop the current log */
+        return DLT_RETURN_LOGGING_DISABLED;
+    }
+
+    /* Compare dlt_user_init_state to INIT_UNITIALIZED. If equal it will be set to INIT_IN_PROGRESS.
+     * Call returns DLT_RETURN_OK init state != INIT_UNITIALIZED
+     * That way it's no problem, if two threads enter this function, because only the very first one will
+     * pass fully. The other one will immediately return, because when it executes the atomic function
+     * dlt_user_init_state won't be INIT_UNITIALIZED anymore.
+     * This is not handled via a simple boolean to prevent issues with shutting down while the init is still running.
+     * Furthermore, this makes sure we enter some function only when dlt_init is fully done.
+     * */
+    enum InitState expectedInitState = INIT_UNITIALIZED;
+    if (!(atomic_compare_exchange_strong(&dlt_user_init_state, &expectedInitState, INIT_IN_PROGRESS))) {
         return DLT_RETURN_OK;
+    }
```

Comment:
Hi @michael-methner the described scenario could leave the library in an initialized state.
Adding a mutex here sounds like a good idea. I'll add it later

Reply:
done :)

Reply:
@michael-methner It's done in case you did not notice it yet :)

Reply:
Hello @alexmohr ,
sorry for the delay and thank you very much for the rework. I was quite busy the last weeks. I need some time to carefully review this as it is related to race conditions ;). I hope I can manage before end of this week.

One first question: Would it make sense to use the same mutex in dlt_free() as well to avoid race conditions between dlt_init() and dlt_free().

Reply:
Thanks for the feedback. Both variables are changed to `_Atomic` and it is building locally with clang 10 and gcc 10.3